internal/beef/beef.go

```diff
@@ -0,0 +1,277 @@
+		if err != nil {
+			return nil, nil, err
+		}
+		hexBytes = remainingBytes
+		bumpPaths = append(bumpPaths, bumpPath)
+	}
+
+	return bumpPaths, hexBytes, nil
+}
+
+func decodeBUMPLevel(nLeaves bt.VarInt, hexBytes []byte) ([]BUMPLeaf, []byte, error) {
+	bumpPath := make([]BUMPLeaf, 0)
+	for i := 0; i < int(nLeaves); i++ {
+		if len(hexBytes) == 0 {
+			return nil, nil, fmt.Errorf("insufficient bytes to extract offset for %d leaf of %d leaves", i, int(nLeaves))
+		}
+
+		offset, bytesUsed := bt.NewVarIntFromBytes(hexBytes)
+		hexBytes = hexBytes[bytesUsed:]
+
+		if len(hexBytes) == 0 {
+			return nil, nil, fmt.Errorf("insufficient bytes to extract flag for %d leaf of %d leaves", i, int(nLeaves))
+		}
+
+		flag := hexBytes[0]
+		hexBytes = hexBytes[1:]
+
+		if flag != dataFlag && flag != duplicateFlag && flag != txIDFlag {
+			return nil, nil, fmt.Errorf("invalid flag: %d for %d leaf of %d leaves", flag, i, int(nLeaves))
+		}
+
+		if flag == duplicateFlag {
+			bumpLeaf := BUMPLeaf{
+				Offset:    uint64(offset),
+				Duplicate: true,
+			}
+			bumpPath = append(bumpPath, bumpLeaf)
+			continue
+		}
+
+		if len(hexBytes) < hashBytesCount {
+			return nil, nil, errors.New("insufficient bytes to extract hash of path")
+		}
+
+		hash := hex.EncodeToString(bt.ReverseBytes(hexBytes[:hashBytesCount]))
+		hexBytes = hexBytes[hashBytesCount:]
+
+		bumpLeaf := BUMPLeaf{
+			Hash:   hash,
+			Offset: uint64(offset),
+		}
+		if flag == txIDFlag {
+			bumpLeaf.TxId = true
+		}
+		bumpPath = append(bumpPath, bumpLeaf)
+	}
+
+	return bumpPath, hexBytes, nil
+}
+
+func decodeTransactionsWithPathIndexes(bytes []byte) ([]*TxData, []byte, error) {
+	if len(bytes) == 0 {
+		return nil, nil, errors.New("invalid BEEF - no transaction")
+	}
+
+	nTransactions, offset := bt.NewVarIntFromBytes(bytes)
+
+	if nTransactions < 2 {
+		return nil, nil, errors.New("invalid BEEF- not enough transactions provided to decode BEEF")
+	}
+
+	bytes = bytes[offset:]
+
+	transactions := make([]*TxData, 0, int(nTransactions))
+
+	for i := 0; i < int(nTransactions); i++ {
+		tx, offset, err := bt.NewTxFromStream(bytes)
+		if err != nil {
+			return nil, nil, err
+		}
+		bytes = bytes[offset:]
+
+		if len(bytes) == 0 {
+			return nil, nil, errors.New("invalid BEEF - no HasBUMP flag")
+		}
+
+		var pathIndex *bt.VarInt
+
+		switch bytes[0] {
+		case HasBump:
+			bytes = bytes[1:]
+			if len(bytes) == 0 {
+				return nil, nil, errors.New("invalid BEEF - HasBUMP flag set, but no BUMP index")
+			}
+			value, offset := bt.NewVarIntFromBytes(bytes)
+			pathIndex = &value
+			bytes = bytes[offset:]
+		case HasNoBump:
+			bytes = bytes[1:]
+		default:
+			return nil, nil, fmt.Errorf("invalid HasCMP flag for transaction at index %d", i)
+		}
+
+		transactions = append(transactions, &TxData{
+			Transaction: tx,
+			BumpIndex:   pathIndex,
+		})
+	}
+
+	return transactions, bytes, nil
+}
+
+func extractBytesWithoutVersionAndMarker(beefBytes []byte) ([]byte, error) {
+	if len(beefBytes) < 4 {
```

Comment:
I believe we can assume that we already have the hex validated and do not need to repeat the validation

Reply:
I'm not sure about that. `CheckBeefFormat()` is currently used for checking the format, because both formats come on the same endpoint. If there will be a situation when we expect BEEF, we may just use DecodeBEEF() without checking the format first, and it should still throw an error and inform us that the tx we passed is not in fact BEEF. 

So IMO, it should stay there. But I refactored it and used the already existing `CheckBeefFormat()`, so we don't repeat ourselves :D